keps/sig-storage/3333-reconcile-default-storage-class/README.md

```diff
@@ -0,0 +1,622 @@
+   no default SC for some time and second case above applies here too.
+   Re-creating the storage class to change parameters can be useful in cases
+   there is a quota set for the SC, and since the quota is coupled with SC name
+   users can not use Option 1 because second SC would have a different name
+   and so the existing quota would not apply to it.
+
+3. Defined ordering during cluster installation. Kubernetes cluster installation
+   tools must be currently smart enough to create a default SC before starting
+   anything that may create PVCs that need it. If such a tool supports multiple
+   cloud providers, storage backends and addons that require storage (such an image
+   registry), it may be quite complicated to do the ordering right.
+
+### Goals
+
+* Loosen ordering requirements between PVCs and default SC creation.
+
+### Non-Goals
+
+* Introduce new API for default SCs.
+
+## Proposal
+
+Change PV controller behavior so it will assign a default SC to unbound PVCs
+with `pvc.spec.storageClassName=nil` after any storage class is marked as
+default.
+
+### User Stories (Optional)
+
+#### Story 1
+
+Admin needs to change the default SC from SC1 to SC2
+1. The admin marks the current default SC1 as non-default.
+2. Another user creates PVC requesting a default SC, by leaving
+   `pvc.spec.storageClassName=nil`. The default SC does not exist at
+   this point, therefore the admission plugin leaves the PVC untouched with
+   `pvc.spec.storageClassName=nil`.
+3. The admin marks SC2 as default.
+4. PV controller, when reconciling the PVC, updates
+   `pvc.spec.storageClassName=nil` to the new SC2.
+5. PV controller uses the new SC2 when binding / provisioning the PVC.
+
+#### Story 2
+
+An installation tool wants to install Kubernetes with a CSI driver providing
+a default SC and an application that wants to use it (such as image registry).
+
+1. The installer creates PVC for the image registry first, requesting the
+   default storage class by leaving `pvc.spec.storageClassName=nil`.
+2. The installer creates a default SC.
+3. PV controller, when reconciling the PVC, updates
+   `pvc.spec.storageClassName=nil` to the new default SC.
+4. PV controller uses the new default SC when binding / provisioning the PVC.
+
+
+#### Story 3 (current behavior)
+
+User wants to provision a volume and there is one default storage class set by
+admin.
+
+1. Admin creates a default storage class.
+2. Another user creates PVC requesting a default SC, by leaving
+   `pvc.spec.storageClassName=nil`. Since the default SC already exists
+   the admission plugin changes the `nil` to a name of the default storage 
+   class.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Behavior change
+
+Currently, when a default SC is not available at PVC creation time, a PVC
+requesting a default SC (`pvc.spec.storageClassName=nil`) will keep
+`pvc.spec.storageClassName=nil` forever. Such PVC can be bound only to PV with
+`pvc.spec.storageClassName=""`.
```

Comment:
@RomanBednar I understand it should be changed to:
Such PVC can be bound only to PV with `pv.spec.storageClassName=""`.